File: abc265/e.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

int main()
{
    ll N, M;
    cin >> N >> M;
    ll A, B, C, D, E, F;
    cin >> A >> B >> C >> D >> E >> F;
    set<pair<ll, ll>> obstructions;
    for (ll i = 0; i < M; i++)
    {
        ll x, y;
        cin >> x >> y;
        obstructions.insert({x, y});
    }
    ll MOD = 998244353;
    map<pair<ll, ll>, ll> dp;
    dp[pair<ll, ll>{0, 0}] = 1;
    vector<pair<ll, ll>> dxy = vector<pair<ll, ll>>{
        {A, B},
        {C, D},
        {E, F},
    };
    for (int i = 0; i < N; i++)
    {
        map<pair<ll, ll>, ll> tmp_dp;

        for (auto [coordinate, cnt_of_path] : dp)
        {
            auto x = coordinate.first;
            auto y = coordinate.second;

            for (auto [dx, dy] : dxy)
            {
                auto n = pair<ll, ll>{x + dx, y + dy};
                auto n_is_not_obstruction = obstructions.find(n) == obstructions.end();
                if (n_is_not_obstruction)
                {
                    tmp_dp[n] += cnt_of_path;
                    tmp_dp[n] %= MOD;
                }
            }
        }
        swap(dp, tmp_dp);
    }
    ll ans = 0;
    for (auto [k, v] : dp)
    {
        ans += v;
        ans %= MOD;
    }
    cout << ans << endl;
}

File: abc265/e.py
# C++で実装し直す。
from collections import defaultdict
N,M = map(int,input().split())
A,B,C,D,E,F = map(int,input().split())

obstructions = set([tuple(map(int,input().split())) for i in range(M)])

dp = [
    defaultdict(int) for i in range(N+1)
]
MOD= 998244353
dp[0][(0,0)] = 1
for i in range(N):
    for coordinate,path_cnt in dp[i].items():
        x,y = coordinate
        for next_coordinates in [(x+A,y+B),(x+C,y+D),(x+E,y+F)]:
            dp[i+1][next_coordinates] = (dp[i][(x,y)] + dp[i+1][next_coordinates]) % MOD

ans = 0
for coordinate,path_cnt in dp[N].items():
    ans = (ans + path_cnt) % MOD
print(ans)